File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/AddEnemiesClosure.java
/*
 * Created on Oct 8, 2005
 */
package rustleund.fightingfantasy.framework.closures.impl;

import java.util.ArrayList;
import java.util.List;

import org.w3c.dom.Element;

import rustleund.fightingfantasy.framework.base.*;
import rustleund.fightingfantasy.framework.closures.Closure;
import rustleund.fightingfantasy.framework.closures.ClosureLoader;

/**
 * @author rustlea
 */
public class AddEnemiesClosure implements Closure {

	private final List<EnemyState> enemiesToAdd;
	private final int battleId;
	private int waitTime;

	public AddEnemiesClosure(Element element, ClosureLoader closureLoader) {
		this.enemiesToAdd = new ArrayList<>();

		Element battleElement = (Element) element.getParentNode();
		while (!battleElement.getLocalName().equals("battle")) {
			battleElement = (Element) battleElement.getParentNode();
		}

		this.battleId = Integer.parseInt(battleElement.getAttribute("id"));
		this.waitTime = Integer.parseInt(element.getAttribute("wait"));

		XMLUtilKt.getChildElementsByName(element, "enemy").iterator()
				.forEachRemaining(enemyTag -> enemiesToAdd.add(new EnemyState(enemyTag, closureLoader)));
	}

	@Override
	public boolean execute(GameState gameState) {
		if (waitTime == 0) {
			BattleState battleState = gameState.getPageState().getBattle(battleId);
			Enemies enemies = battleState.getEnemies();
			for (EnemyState enemy : enemiesToAdd) {
				enemies.addEnemy(enemy);
			}
		}
		waitTime--;
		return true;
	}

}

File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/AddEnemiesClosure.kt
/*
 * Created on Oct 8, 2005
 */
package rustleund.fightingfantasy.framework.closures.impl

import org.w3c.dom.Element
import rustleund.fightingfantasy.framework.base.*
import rustleund.fightingfantasy.framework.closures.ClosureLoader
import rustleund.fightingfantasy.framework.closures.Closure

class AddEnemiesClosure(element: Element, closureLoader: ClosureLoader) : Closure {

    private val enemiesToAdd = getChildElementsByName(element, "enemy")
        .map { EnemyState(it, closureLoader) }.toList()
    private val battleId = element.getAncestorByName("battle")?.optionalIntAttribute("id")
        ?: throw IllegalArgumentException("No ancestor battle element or ID found")

    private var waitTime = element.getAttribute("wait").toInt()

    override fun execute(gameState: GameState): Boolean {
        if (waitTime == 0) {
            val battleState = gameState.pageState.getBattle(battleId)
            enemiesToAdd.forEach { battleState.enemies.addEnemy(it) }
        }
        waitTime--
        return true
    }
}